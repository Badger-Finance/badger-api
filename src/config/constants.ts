import { ethers } from 'ethers';

const RAW_TOKENS = {
  // eth tokens
  BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
  DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',
  WBTC: '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599',
  WETH: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
  SUSHI: '0x6b3595068778dd592e39a122f4f5a5cf09c90fe2',

  // bsc tokens
  CAKE: '0x0E09FaBB73Bd3Ade0a17ECC321fD13a19e81cE82',
  WBNB: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
  BTCB: '0x7130d2A12B9BCbFAe4f2634d864A1Ee1Ce3Ead9c',
  BSC_BADGER: '0x753fbc5800a8C8e3Fb6DC6415810d627A387Dfc9',
  BSC_BBADGER: '0x1F7216fdB338247512Ec99715587bb97BBf96eae',
  BSC_BDIGG: '0x5986D5c77c65e5801a5cAa4fAE80089f870A71dA',

  // curve tokens
  CRV_RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
  CRV_TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
  CRV_SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',

  // uniswap tokens
  UNI_BADGER_WBTC: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
  UNI_DIGG_WBTC: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

  // sushiswap tokens
  SUSHI_ETH_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
  SUSHI_BADGER_WBTC: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
  SUSHI_DIGG_WBTC: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',

  // pancakeswap tokens
  PANCAKE_BNB_BTCB: '0x356b7d0d3c54F22C82B7a670C6Ba9E2381b0624c',
  PANCAKE_BBADGER_BTCB: '0x87Ae7b5c43D4e160cDB9427a78BA87B9503ee37b',
  PANCAKE_BDIGG_BTCB: '0xfbfa92e037e37F946c0105902640914E3aCe6752',

  // eth vault tokens
  BBADGER: '0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28',
  BDIGG: '0x7e7e112a68d8d2e221e11047a72ffc1065c38e1a',
  BUNI_BADGER_WBTC: '0x235c9e24d3fb2fafd58a2e49d454fdcd2dbf7ff1',
  BUNI_DIGG_WBTC: '0xc17078fdd324cc473f8175dc5290fae5f2e84714',
  BSUSHI_ETH_WBTC: '0x758a43ee2bff8230eeb784879cdcff4828f2544d',
  BSUSHI_BADGER_WBTC: '0x1862a18181346ebd9edaf800804f89190def24a5',
  BSUSHI_DIGG_WBTC: '0x88128580acdd9c04ce47afce196875747bf2a9f6',
  BCRV_SBTC: '0xd04c48a53c111300ad41190d63681ed3dad998ec',
  BCRV_RENBTC: '0x6def55d2e18486b9ddfaa075bc4e4ee0b28c1545',
  BCRV_TBTC: '0xb9d076fde463dbc9f915e5392f807315bf940334',
  BCRV_HRENBTC: '0xaf5a1decfa95baf63e0084a35c62592b774a2a87',
  BVYWBTC: '0x4b92d19c11435614cd49af1b589001b7c08cd4d5',

  // bsc vault tokens
  BPANCAKE_BNB_BTCB: '0xaf4B9C4b545D5324904bAa15e29796D2E2f90813',
  BPANCAKE_BBADGER_BTCB: '0x857F91f735f4B03b19D2b5c6E476C73DB8241F55',
  BPANCAKE_BDIGG_BTCB: '0xa861Ba302674b08f7F2F24381b705870521DDfed',
};

export const TOKENS = Object.fromEntries(
  Object.entries(RAW_TOKENS).map(([key, val]) => [key, ethers.utils.getAddress(val)]),
);

export enum Protocol {
  Curve = 'curve',
  Sushiswap = 'sushiswap',
  Uniswap = 'uniswap',
  Pancakeswap = 'pancakeswap',
  Yearn = 'yearn',
}

// data point constants - index two times per hour, 48 per day
export const CURRENT = 0;
export const ONE_DAY = 24 * 2;
export const ONE_MINUTE_MS = 60 * 1000;
export const THREE_DAYS = ONE_DAY * 3;
export const SEVEN_DAYS = ONE_DAY * 7;
export const THIRTY_DAYS = ONE_DAY * 30;
export const SAMPLE_DAYS = THIRTY_DAYS + 1;
export const ONE_YEAR_MS = 365 * 24 * 60 * 60 * 1000;

// data access constants
export const APY_SNAPSHOTS_DATA = process.env.APY_SNAPSHOTS_DATA || 'MISSING REQUIRED ENV VAR';
export const ASSET_DATA = process.env.ASSET_DATA || 'MISSING REQUIRED ENV VAR';
export const PRICE_DATA = process.env.PRICE_DATA || 'MISSING REQUIRED ENV VAR';
export const SETT_SNAPSHOTS_DATA = process.env.SETT_SNAPSHOTS_DATA || 'MISSING REQUIRED ENV VAR';
export const REWARD_DATA = process.env.REWARD_DATA || 'MISSING REQUIRED ENV VAR';
export const UNISWAP_URL = process.env.UNISWAP || 'MISSING REQUIRED ENV VAR';
export const SUSHISWAP_URL = process.env.SUSHISWAP || 'MISSING REQUIRED ENV VAR';
export const PANCAKESWAP_URL = process.env.PANCAKESWAP || 'MISSING REQUIRED ENV VAR';
export const BADGER_URL = process.env.BADGER || 'MISSING REQUIRED ENV VAR';
export const BADGER_BSC_URL = process.env.BADGER_BSC || 'MISSING REQUIRED ENV VAR';
export const BADGER_DAO_URL = process.env.BADGER_DAO || 'MISSING REQUIRED ENV VAR';
export const MASTERCHEF_URL = process.env.MASTERCHEF || 'MISSING REQUIRED ENV VAR';
export const BOUNCER_PROOFS = process.env.BOUNCER_PROOFS || 'MISSING REQUIRED ENV VAR';

// third party api constants
export const CURVE_API_URL = 'https://stats.curve.fi/raw-stats/apys.json';
export const COINGECKO_URL = 'https://api.coingecko.com/api/v3/simple';

export enum Provider {
  Cloudflare = 'https://cloudflare-eth.com/',
  Binance = 'https://bsc-dataseed.binance.org/',
  MyEtherWallet = 'https://nodes.mewapi.io/rpc/eth',
}
export const SUSHI_CHEF = '0xc2EdaD668740f1aA35E4D8f227fB8E17dcA888Cd';
export const PANCAKE_CHEF = '0x73feaa1eE314F8c655E354234017bE2193C9E24E';
