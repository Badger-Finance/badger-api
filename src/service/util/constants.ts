export const TOKENS = {
	// general tokens
	RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
	BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
	TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
	SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',
	DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',

	// uniswap tokens
	UNI_BADGER: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
	UNI_DIGG: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

	// sushiswap tokens
	SUSHI_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
	SUSHI_BADGER: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
	SUSHI_DIGG: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',
};
