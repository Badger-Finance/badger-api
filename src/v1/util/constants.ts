import { ethers } from 'ethers';

export const TOKENS = {
  // general tokens
  RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
  BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
  TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
  SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',
  DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',

  // uniswap tokens
  UNI_BADGER: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
  UNI_DIGG: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

  // sushiswap tokens
  SUSHI_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
  SUSHI_BADGER: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
  SUSHI_DIGG: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',
};

// data point constants - index two times per hour, 48 per day
export const CURRENT = 0;
export const ONE_DAY = 24 * 2;
export const THREE_DAYS = ONE_DAY * 3;
export const SEVEN_DAYS = ONE_DAY * 7;
export const THIRTY_DAYS = ONE_DAY * 30;
export const SAMPLE_DAYS = THIRTY_DAYS + 1;
export const ONE_YEAR_MS = 365 * 24 * 60 * 60 * 1000;

// data access constants
export const ASSET_DATA = process.env.ASSET_DATA || 'MISSING REQUIRED ENV VAR';
export const UNISWAP_URL = process.env.UNISWAP || 'MISSING REQUIRED ENV VAR';
export const SUSHISWAP_URL = process.env.SUSHISWAP || 'MISSING REQUIRED ENV VAR';
export const BADGER_URL = process.env.BADGER || 'MISSING REQUIRED ENV VAR';
export const BADGER_DAO_URL = process.env.BADGER_DAO_URL || 'MISSING REQUIRED ENV VAR';
export const MASTERCHEF_URL = process.env.MASTERCHEF || 'MISSING REQUIRED ENV VAR';

// third party api constants
export const CURVE_API_URL = 'https://stats.curve.fi/raw-stats/apys.json';
export const ETHERS_JSONRPC_PROVIDER = new ethers.providers.JsonRpcProvider('https://cloudflare-eth.com/');
