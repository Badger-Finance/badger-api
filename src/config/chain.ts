import { ethers } from 'ethers';
import { SettDefinition } from '../interface/Sett';
import { Token } from '../interface/Token';
import { Provider } from './constants';
import { ethSetts } from './setts/eth-setts';

export interface Chain {
  readonly name: string;
  readonly symbol: string;
  readonly chainId: string;
  readonly tokens: Token[];
  readonly setts: SettDefinition[];
  readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
  public name = 'Ethereum';
  public symbol = 'eth';
  public chainId = '0x01';
  public tokens = [];
  public setts = ethSetts;
  public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
  public name = 'BinanceSmartChain';
  public symbol = 'bsc';
  public chainId = '0x61';
  public tokens = [];
  public setts = [];
  public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 * TODO: Should not supported throw a bad request instead?
 */
export const resolveChainQuery = (symbol: string): Chain => {
  if (!symbol) return eth;
  const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
  if (!chain) return eth;
  return chain;
};
