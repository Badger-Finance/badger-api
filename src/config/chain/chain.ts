import { BadRequest } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { SettDefinition } from '../../interface/Sett';
import { Token } from '../../interface/Token';
import { Provider } from '../constants';
import { bscSetts } from './bsc';
import { ethSetts } from './eth';

export interface Chain {
	readonly name: string;
	readonly symbol: string;
	readonly chainId: string;
	readonly tokens: Token[];
	readonly setts: SettDefinition[];
	readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
	public name = 'Ethereum';
	public symbol = 'eth';
	public chainId = '0x01';
	public tokens = [];
	public setts = ethSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
	public name = 'BinanceSmartChain';
	public symbol = 'bsc';
	public chainId = '0x61';
	public tokens = [];
	public setts = bscSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 */
export const resolveChainQuery = (symbol: string): Chain => {
	if (!symbol) return eth;
	const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
	if (!chain) throw new BadRequest('Invalid chain');
	return chain;
};
