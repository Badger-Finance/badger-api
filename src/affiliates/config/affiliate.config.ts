import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
