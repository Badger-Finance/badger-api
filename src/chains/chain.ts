import { Arbitrum } from './config/arbitrum.config';
import { BinanceSmartChain } from './config/bsc.config';
import { Chain } from './config/chain.config';
import { Ethereum } from './config/eth.config';
import { Polygon } from './config/matic.config';

/**
 * Instantiate objects for registration.
 */
export const loadChains = (): Chain[] => {
  return [new Ethereum(), new BinanceSmartChain(), new Polygon(), new Arbitrum()];
};
