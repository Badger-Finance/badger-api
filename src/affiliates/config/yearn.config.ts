import { BigNumber, ethers } from 'ethers';
import { Chain } from '../../chains/config/chain.config';
import { guestListAbi } from '../../config/abi/guest-list.abi';
import { yearnAffiliateVaultWrapperAbi } from '../../config/abi/yearn-affiliate-vault-wrapper.abi';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { getToken } from '../../tokens/tokens.utils';
import { Affiliate } from './affiliate.config';

export class Yearn extends Affiliate {
  constructor() {
    super(Protocol.Yearn);
    Affiliate.register(Protocol.Yearn, this);
  }

  async getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData> {
    const affiliateData: SettAffiliateData = {
      protocol: Protocol.Yearn,
    };
    if (sett.hasBouncer) {
      try {
        const vaultWrapper = new ethers.Contract(sett.settToken, yearnAffiliateVaultWrapperAbi, chain.provider);
        const guestListAddress = await vaultWrapper.guestList();
        const guesList = new ethers.Contract(guestListAddress, guestListAbi, chain.provider);
        const [available, limit]: [BigNumber, BigNumber] = await Promise.all([
          guesList.remainingTotalDepositAllowed(),
          guesList.totalDepositCap(),
        ]);
        const depositToken = getToken(sett.depositToken);
        affiliateData.availableDepositLimit = parseFloat(ethers.utils.formatUnits(available, depositToken.decimals));
        affiliateData.depositLimit = parseFloat(ethers.utils.formatUnits(limit, depositToken.decimals));
      } catch {
        affiliateData.availableDepositLimit = 0;
        affiliateData.depositLimit = 50;
      }
    }
    return affiliateData;
  }
}
