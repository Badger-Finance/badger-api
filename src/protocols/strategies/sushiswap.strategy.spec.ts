import { Network } from '@badger-dao/sdk';

import { Chain } from '../../chains/config/chain.config';
import { TOKENS } from '../../config/tokens.config';
import * as sushiswapGraph from '../../graphql/generated/sushiswap';
import { MOCK_VAULT_DEFINITION } from '../../test/constants';
import { setupMockChain } from '../../test/mocks.utils';
import { getSushiswapYieldSources } from './sushiswap.strategy';

describe('sushiswap.strategy', () => {
  let chain: Chain;

  beforeEach(() => {
    chain = setupMockChain();
  });

  describe('getSushiswapYieldSources', () => {
    it('returns no value sources given no trade data in the subgraph', async () => {
      jest.spyOn(sushiswapGraph, 'getSdk').mockImplementation((_client) => {
        return {
          SushiPairDayDatas(_v, _r): Promise<sushiswapGraph.SushiPairDayDatasQuery> {
            return Promise.resolve({ pairDayDatas: [] });
          },
        };
      });
      const result = await getSushiswapYieldSources(chain, MOCK_VAULT_DEFINITION);
      expect(result).toMatchObject([]);
    });

    it.each([[Network.Ethereum], [Network.Polygon], [Network.Arbitrum]])(
      'returns trade fees for the pair given trade data in the subgraph for %s',
      async (network) => {
        chain = setupMockChain({ network });
        jest.spyOn(sushiswapGraph, 'getSdk').mockImplementation((_client) => {
          return {
            SushiPairDayDatas(_v, _r): Promise<sushiswapGraph.SushiPairDayDatasQuery> {
              return Promise.resolve({
                pairDayDatas: [
                  {
                    token0: {
                      id: TOKENS.BADGER,
                    },
                    token1: {
                      id: TOKENS.WBTC,
                    },
                    reserve0: 5000,
                    reserve1: 1,
                    reserveUSD: 40000,
                    volumeUSD: 2500000,
                    date: 85000,
                  },
                ],
              });
            },
          };
        });
        const result = await getSushiswapYieldSources(chain, MOCK_VAULT_DEFINITION);
        expect(result).toMatchSnapshot();
      },
    );
  });
});
