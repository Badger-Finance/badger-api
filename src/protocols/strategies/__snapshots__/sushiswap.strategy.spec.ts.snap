// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sushiswap.strategy getSushiswapYieldSources returns trade fees for the pair given trade data in the subgraph for arbitrum 1`] = `
Array [
  CachedYieldSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_badger_lp_fees_flat",
    "name": "Badger LP Fees",
    "performance": Object {
      "baseYield": 5703.125,
      "grossYield": 5703.125,
      "maxGrossYield": 5703.125,
      "maxYield": 5703.125,
      "minGrossYield": 5703.125,
      "minYield": 5703.125,
    },
    "type": "lp_fee",
  },
]
`;

exports[`sushiswap.strategy getSushiswapYieldSources returns trade fees for the pair given trade data in the subgraph for ethereum 1`] = `
Array [
  CachedYieldSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_badger_lp_fees_flat",
    "name": "Badger LP Fees",
    "performance": Object {
      "baseYield": 5703.125,
      "grossYield": 5703.125,
      "maxGrossYield": 5703.125,
      "maxYield": 5703.125,
      "minGrossYield": 5703.125,
      "minYield": 5703.125,
    },
    "type": "lp_fee",
  },
]
`;

exports[`sushiswap.strategy getSushiswapYieldSources returns trade fees for the pair given trade data in the subgraph for polygon 1`] = `
Array [
  CachedYieldSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_badger_lp_fees_flat",
    "name": "Badger LP Fees",
    "performance": Object {
      "baseYield": 5703.125,
      "grossYield": 5703.125,
      "maxGrossYield": 5703.125,
      "maxYield": 5703.125,
      "minGrossYield": 5703.125,
      "minYield": 5703.125,
    },
    "type": "lp_fee",
  },
]
`;
